---
title: <span style="font-family:Georgia;">Integration of mutational and expression data with DIABLO</span>
author: "Víctor Jiménez Martínez"
date: "2022"
output: 
  html_document:
    toc: true
    theme: united
---

## Introduction

Data Integration Analysis for Biomarker discovery using Latent component method for Omics studies (DIABLO)^1^ is the multi-block discriminant analysis from the mixOmics^2^ R package. DIABLO allows the integration of several omics data sets on a set of samples to identify a multi-omics signature that explains a given outcome. DIABLO applies a sparse General-ized Canonical Correlation Analysis (sGCCA)^3^, a multivariate dimension reduction methodology that uses singular value decomposition to select correlated variables from several omics data sets by maximizing the covariance between linear combinations of variables and projecting the data into smaller dimensional subspace spanned by the components. This way, highly correlated variables between and within omics datasets are identified and integrated in the model to explain response to treatment in pre-treatment samples.

In this case, the layers integrated were expression data, functional SNVs and INDELs mutation data and MCPCounter estimated relative abundances of immune cell populations expected to identify a correlated signature discriminating by response. Before applying the method data must be filtered and prepared. TPM normalized counts were used for expression data, while z-score was applied to deconvolution data and gene-length normalized mutation data. Regarding the filtering, it is mainly recommended for computing and R memory purposes and for facilitating data mining (assuming that out of thousands of variables, most of them will not be useful to explain a biological outcome). The filtering is recommended to be based on the variance across all the samples (the ones holding larger variances are more likely to explain differences). Usually, no more than 5.000-10.000 features are recommended. 

Another factor to consider is the number of samples available. A total of 21 pre-treatment samples with WES and RNAseq data were available: 14 non-responders and 7 responders. With the purpose of building a model that allowed to discriminate, 5 samples of each group were randomly se-lected to build the model, while the rest were kept to further test the model with “external” data.

## Libraries

Libraries needed for the analyses are loaded below.

```{r, results="hide", message=FALSE, warning=FALSE}
# R version used in this script is 4.2.0
# library(devtools)
# install_github("mixOmicsTeam/mixOmics") # I had to install the latest version
library(ggplot2) # plots
library(EDASeq) # for RLE plots
library(mixOmics) # for integrating omics data through multivariate methods
library(tidyr) # to manage data
library(dplyr) # to manage data
library(biomaRt)
library(ggpubr)
library(edgeR)
```

# Loading and preparing data

```{r}
# MCPCounter results
mcp <- read.delim("/omics/groups/OE0436/data/jimenez/outputs/rnaseq/deconvolution/mcp_decon_immune_pre.tsv")
mcp <- t(mcp[,-29]) # features must be rows

# TPM Expression data
tpm <- read.delim("/omics/groups/OE0436/data/jimenez/data/rnaseq/rnaseq_tpm_met1_protein_coding.tsv")
colnames(tpm)[1:5] <- sapply(colnames(tpm)[1:5], function(x) gsub("^X", "", x)) # adapt five first PIDs

# SNV data
snv <- read.delim("/omics/groups/OE0436/data/jimenez/data/snv_indels/vcf/all_snvs_all_patients_metastasis1_definitive.vcf")  

# Filter rows that do not represent a somatic mutation and select those interesting fields
snv_select <- snv %>% 
  filter(RECLASSIFICATION=="somatic" | RECLASSIFICATION=="somatic_precious") %>%
  dplyr::select(PID, GENE, ANNOVAR_FUNCTION, EXONIC_CLASSIFICATION, X.CHROM, POS)

# Add mutation level
snv_select$mut_level <- paste(snv_select$ANNOVAR_FUNCTION, snv_select$EXONIC_CLASSIFICATION, sep = "_")
snv_select$mut_level = gsub("_\\.", "", snv_select$mut_level)
snv_select$mut_level = gsub(" ", "_", gsub(";", "_", snv_select$mut_level))

# Modify mutation levels
snv_select$mut_level[snv_select$mut_level == "exonic_stopgain"] <- "stopgain"
snv_select$mut_level[snv_select$mut_level == "exonic_stoploss"] <- "stoploss"
snv_select$mut_level[snv_select$mut_level == "exonic_nonsynonymous_SNV"] <- "nonsynonymous SNV"
snv_select$mut_level[snv_select$mut_level == "exonic_splicing_unknown" |
                       snv_select$mut_level == "exonic_splicing_nonsynonymous_SNV" |
                       snv_select$mut_level == "exonic_splicing_synonymous_SNV"] <- "splicing"

# INDEL data 
indel <- read.delim("/omics/groups/OE0436/data/jimenez/data/snv_indels/vcf/all_indels_all_patients_metastasis1_definitive.vcf") 

# Filter rows that do not represent a somatic mutation and select those interesting fields
indel_select <- indel %>%
  dplyr::select(PID, GENE, ANNOVAR_FUNCTION, EXONIC_CLASSIFICATION, X.CHROM, POS)

# Add mutation level
indel_select$mut_level <- paste(indel_select$ANNOVAR_FUNCTION, indel_select$EXONIC_CLASSIFICATION, sep = "_")
indel_select$mut_level = gsub("_\\.", "", indel_select$mut_level)
indel_select$mut_level = gsub(" ", "_", gsub(";", "_", indel_select$mut_level))

# Modify mutation levels
indel_select$mut_level[indel_select$mut_level == "exonic_frameshift_deletion"] <- "frameshift deletion"
indel_select$mut_level[indel_select$mut_level == "exonic_frameshift_insertion"] <- "frameshift insertion"
indel_select$mut_level[indel_select$mut_level == "exonic_nonframeshift_deletion"] <- "nonframeshift deletion"
indel_select$mut_level[indel_select$mut_level == "exonic_nonframeshift_insertion"] <- "nonframeshift insertion"
indel_select$mut_level[indel_select$mut_level == "exonic_unknown"] <- "unknown"
indel_select$mut_level[indel_select$mut_level == "exonic_stopgain"] <- "stopgain"

# Combine snvs and indels
snv_indels <- rbind(snv_select, indel_select)

# Remove exon information from "splicing" SNVs
snv_indels$GENE = gsub("\\(.*", "", snv_indels$GENE)

# Construct a matrix of mutations
snv_indels <- snv_indels %>% dplyr::select(PID, GENE) %>%
  mutate(n = 1) %>% dplyr::distinct() %>% tidyr::spread(GENE, n)
rownames(snv_indels) <- snv_indels$PID
snv_indels <- snv_indels %>% dplyr::select(-PID) %>% t()
snv_indels[is.na(snv_indels)] <- 0
```

All our available data in pre-treatment samples has information for the same patients: 21 out of the 24 in the cohort. We have to order them in the same order and read pheno data accordingly. 

```{r}
# Check order
identical(colnames(tpm), colnames(mcp))
identical(colnames(tpm), colnames(mcp))

# Pheno data with information on treatment and response
pheno <- read.csv(file = "/omics/groups/OE0436/data/jimenez/data/PID_treatment.csv", 
                  header = TRUE, sep = ";", row.names = "PID")

colnames(pheno) = c("therapy", "evolution") # change names

pheno <- pheno %>% 
  mutate(response = ifelse(evolution == "PR", "responders", "non-responders"))

pheno <- pheno[colnames(tpm), ]
```

Since normalization will be tested, let's filter data.

```{r}
# Filter TPM
tpm <- tpm[rowSums(tpm) > 10, ]

cat(dim(tpm)[1], 
    "genes with more than 10 counts")

# Keep genes that have counts in at least 80% of each group
# Compute number of samples per experimental group and minimum percent
n.samplespergroup <- table(pheno$response) 
percent <- 0.8 # define the 80%
samples.min <- round(n.samplespergroup *(1-percent), 0) # number of samples that represents the 20%, the maximum amount of 0 allowed


# Data frame with the number of 0 in each gene and sample
pre_0 <- data.frame(row.names = rownames(tpm), 
                   "non-responders" = rep(0, length(rownames(tpm))), 
                   "responders" = rep(0, length(rownames(tpm))))

for (i in 1:2){
  c <- tpm[ , which(colnames(tpm) %in% rownames(pheno[pheno$response == unique(pheno$response)[i], ]))] # subset counts per group
  how_many_0 <- apply(c, 1, function(x) sum(x == 0)) # compute how many 0 per gene
  pre_0[, i] <- as.vector(how_many_0)
}

# Get genes that have counts in at least 80% of the samples of at least one group
# Substract the 80% computed for each group to the counts found for each gene, if it is > 0 it means that at least 80% of samples counted that gene
resta <- apply(pre_0, 1,  function(x) {any(x < as.vector(samples.min))}) 

idx_80 <- rownames(tpm)[which(resta == TRUE)]

cat(length(idx_80), "out of", dim(tpm)[1], 
    "genes counted in 80% of the samples of at least one group")

tpm <- tpm[idx_80, ]

# Let's filter functional 
# At least one mutation in either of one groups
snv_indels <- snv_indels[rowSums(snv_indels[, which(colnames(snv_indels) %in% rownames(pheno)[pheno$response == "responders"])]) > 1 | rowSums(snv_indels[, which(colnames(snv_indels) %in% rownames(pheno)[pheno$response == "non-responders"])]) > 1, ]

dim(snv_indels)

```

## Normalization

With filtered and prepared data, mutation data must be transformed to continous data and normalization of our three layers of information is performed. For mutations, we will try to normalize according to gene length. 

```{r}
################### MUTATION DATA
# Apparently, there are two genes in our snv_indels table in which there are two genes separated by a comma. We keep one of them 
rownames(snv_indels)[c(47,224)] <- unlist(lapply(strsplit(rownames(snv_indels)[c(47,224)], ","), function(x) x[1]))

# First of all, we get the ensembl ID of our genes from the raw file from RNA-seq data
exrna <- read.delim("/omics/groups/OE0436/internal/eugenie/data/k28a/rna_sequencing/rnaseq_counts/metastasis1_K28A-1CNM9W.fpkm_tpm.featureCounts.tsv")

sum(rownames(snv_indels) %in% exrna$name) # all genes have their ensembl id
exrna <- exrna[exrna$name %in% rownames(snv_indels), ] 
which(duplicated(exrna$name)) # there are two duplicated genes: CFB and OC90
exrna <- exrna[-c(137, 185), ]

ensembl_id <- exrna$gene_id[which(exrna$name %in% rownames(snv_indels))] # list of ensembl_id

# Compute gene length using biomart
ensembl <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl") 

keytypes(ensembl) # ensembl_gene_id is what we are providing
searchAttributes(mart = ensembl, pattern = "type") # gene_biotype is what we want to know
searchFilters(mart = ensembl)

# Query
gene_position <- getBM(mart = ensembl, 
                    values = unlist(lapply(strsplit(ensembl_id, "\\."), function(x) x[1])),
                    attributes = c("ensembl_gene_id", "start_position", "end_position"),
                    filters = c("ensembl_gene_id"))# keytype = "ensembl_gene_id")

# One gene has not been found: "ENSG00000107623"
gene_position[371, ] <- c("ENSG00000107623", 1, 1)

rownames(gene_position) <- gene_position$ensembl_gene_id
gene_position <- gene_position[unlist(lapply(strsplit(ensembl_id, "\\."), function(x) x[1])), ]
gene_position$length <- as.numeric(gene_position$end_position) -
  as.numeric(gene_position$start_position)
gene_position["ENSG00000107623", 4] <- 13382 # manually insert gene length (obtained from https://www.bcgsc.ca/)

# Each row is divided by its correspondant gene length 
snv_indels2 <- (snv_indels / gene_position$length) * 1000000 # megabases

# Z-score of this gene length normalized data
zmut <- (snv_indels2 - apply(snv_indels2, 2, mean)) / apply(snv_indels2, 2, sd)
  
# Let's check 
plotRLE(as.matrix(zmut), outline = FALSE, ylim = c(-2, 2), main = "Z-score length normalized mutations (SNV + INDELS)") # variation seems ok
plotRLE(as.matrix(snv_indels), outline = FALSE, ylim = c(-2, 2), main = "Raw mutations (SNV + INDELS)") 

# This kind of plot seems ok
MbyTPM <- log2(zmut[, 5] - zmut[, 20])
hist(as.numeric(MbyTPM), xlab = "log2-ratio", main = "z-score length normalized mutations")
abline(v = 0, col = "red")

################### DECONVOLUTION DATA
# Z-score normalization
zmcp <- (mcp - apply(mcp, 2, mean)) / apply(mcp, 2, sd)
logmcp <- log2(mcp)

# Check 
plotRLE(as.matrix(zmcp), outline = FALSE, ylim = c(-2, 2), main = "Zscore") # it seems controlled

MbyTPM <- log2(zmcp[, 1] - zmcp[,15])
hist(as.numeric(MbyTPM), xlab = "log2-ratio", main = "Differences in z-score MCPCounter estimated relative abundance across samples")
abline(v = 0, col = "red")

# Normality and dispersion are not fully accomplished but a tendency can be seen


################### RNA-seq data
# As already seen in the rnaseq_DEA.Rmd script, our TPM normalized data seems fine
```

The mutation data normalization has been hard, since most of the values are 0 (missing mutations). mixOmics needs continous data, so the z-score transformation with the normalization per gene length should suffice, although we should not expect the data to be completely normal. 

## PCA

Let's again compute the PCA to see the clustering of our data according to the groups across the different layers of information. 

```{r}
############# RNAseq data
# Compute PCA
pca_data <- prcomp(t(tpm), retx=TRUE, center=TRUE, scale=FALSE)
summary(pca_data)$importance[,c(1,2)]

# Scores data
scores <- as.data.frame(pca_data$x[, c("PC1", "PC2")]) # loading coeficients
scores$Response <- pheno$response # response to treatment 
scores$batch <- pheno$batch

# Plot
pdf("/omics/groups/OE0436/data/jimenez/outputs/rnaseq/DEA/protein_coding_genes/PCA_expression_tpm_protein_coding.pdf")

scores.plot <- ggplot(data = scores, 
                      aes(x = PC1, 
                          y = PC2, 
                          colour = Response, 
                          label = rownames(scores))) + 
  geom_point(alpha = I(0.7), size = 4) + 
  geom_hline(yintercept = 0)+
  geom_vline(xintercept = 0)+
  xlab(paste0("PC1 (", round(summary(pca_data)$importance[2,1], 2) * 100, "%)"))+
  ylab(paste0("PC2 (", round(summary(pca_data)$importance[2,2], 2) * 100, "%)"))+
  stat_ellipse() + scale_color_manual(values = c("red", "green")) +
  theme_bw() + ggtitle("PCA with expression data (TPM counts)")

scores.plot

dev.off()

############# Deconvolution data
# Compute PCA
pca_data <- prcomp(t(zmcp), retx=TRUE, center=TRUE, scale=FALSE)
summary(pca_data)$importance[,c(1,2)]

# Scores data
scores <- as.data.frame(pca_data$x[, c("PC1", "PC2")]) # loading coeficients
scores$Response <- pheno$response # response to treatment 
scores$batch <- pheno$batch

# Plot
pdf("/omics/groups/OE0436/data/jimenez/outputs/rnaseq/DEA/protein_coding_genes/PCA_expression_tpm_protein_coding.pdf")

scores.plot <- ggplot(data = scores, 
                      aes(x = PC1, 
                          y = PC2, 
                          colour = Response, 
                          label = rownames(scores))) + 
  geom_point(alpha = I(0.7), size = 4) + 
  geom_hline(yintercept = 0)+
  geom_vline(xintercept = 0)+
  xlab(paste0("PC1 (", round(summary(pca_data)$importance[2,1], 2) * 100, "%)"))+
  ylab(paste0("PC2 (", round(summary(pca_data)$importance[2,2], 2) * 100, "%)"))+
  stat_ellipse() + scale_color_manual(values = c("red", "green")) +
  theme_bw() + ggtitle("PCA with MCPCounter deconvolution data ")

scores.plot

dev.off()

############# Mutation data
# Compute PCA
pca_data <- prcomp(t(zmut), retx=TRUE, center=TRUE, scale=FALSE)
summary(pca_data)$importance[,c(1,2)]

# Scores data
scores <- as.data.frame(pca_data$x[, c("PC1", "PC2")]) # loading coeficients
scores$Response <- pheno$response # response to treatment 
scores$batch <- pheno$batch

# Plot
pdf("/omics/groups/OE0436/data/jimenez/outputs/rnaseq/DEA/protein_coding_genes/PCA_expression_tpm_protein_coding.pdf")

scores.plot <- ggplot(data = scores, 
                      aes(x = PC1, 
                          y = PC2, 
                          colour = Response, 
                          label = rownames(scores))) + 
  geom_point(alpha = I(0.7), size = 4) + 
  geom_hline(yintercept = 0)+
  geom_vline(xintercept = 0)+
  xlab(paste0("PC1 (", round(summary(pca_data)$importance[2,1], 2) * 100, "%)"))+
  ylab(paste0("PC2 (", round(summary(pca_data)$importance[2,2], 2) * 100, "%)"))+
  stat_ellipse() + scale_color_manual(values = c("red", "green")) +
  theme_bw() + ggtitle("PCA with pre-treatment functional mutational (SNV + indels) data ")

scores.plot

dev.off()
```

## Omics data integration through N-block approaches: mixOmics DIABLO

DIABLO, as explained, allows the integration of several omics data sets on a set of samples to identify a multi-omics signature that explains a given outcome. In this case, the integration of expression, deconvolution and mutational data sets on the 21 pre-treatment samples is expected to identify a correlated signature that discriminates by response.

### Filtering

Before applying the method, data is filtered and prepared. Regarding the filtering, it is mainly recommended for computing purposes (tuning the DIABLO model takes a long time and R might not be able to handle so much memory) and to facilitate data mining (out of the thousands of variables, many of them will not be useful to explain the biological outcome). The filtering is recommended to be based on the variance across all the samples: those variables that hold a larger variance are more likely to explain the differences. Usually, no more than the top most variable 5,000-10,000 features are kept. In this analysis, all the estimated cell populations (28) and all the mutations (371) will be used. As for the expression data, the expression for the differentially expressed genes (as obtained in rnaseq_DEA.Rmd) will be used. 

```{r}
# DEG are loaded 
deg <- read.delim("/omics/groups/OE0436/data/jimenez/outputs/rnaseq/DEA/protein_coding_genes/DEG_protein_coding_pre.tsv")

# Read again tpm
tpm <- read.delim("/omics/groups/OE0436/data/jimenez/data/rnaseq/rnaseq_tpm_met1_protein_coding.tsv")
colnames(tpm)[1:5] <- sapply(colnames(tpm)[1:5], function(x) gsub("^X", "", x)) # adapt five first PIDs

tpm_deg <- tpm[rownames(deg), ] # subset for our genes

# We select 5 random samples of both groups to build the model
set.seed(123)
cases_model_idx <- c(sample(which(pheno$response == "responders"), 5),
                     sample(which(pheno$response == "non-responders"), 5))

# And subset our data
zmcp_model <- zmcp[, cases_model_idx]
tpm_deg_model <- tpm_deg[, cases_model_idx]
zmut_model <- zmut[, cases_model_idx]

# Transpose data: samples in rows and variables (genes and probes) in columns
zmcp_model_t <- t(zmcp_model)
tpm_deg_model_t <- t(tpm_deg_model)
zmut_model_t <- t(zmut_model)

# Condition: aged vs young
pheno_model <- pheno[cases_model_idx, ]
cond <- pheno$response[cases_model_idx]
```

### Building the model

Once data is prepared, method may be applied. First of all, a list with all data and a design matrix containing the expected correlation for our omics data are built. For a first iteration without previous knowledge on potential correlations, 0.1 is employed.

```{r}
# Let's explore correlation 
# Build data frame with median for each layer
correlation <- data.frame(row.names = rownames(tpm_deg_model_t),
                          "mRNA" = as.numeric(apply(tpm_deg_model_t, 1, median)),
                          "Deconvolution" = as.numeric(apply(zmcp_model_t, 1, median)),
                          "WES" = as.numeric(apply(zmut_model_t, 1, median)))

# Do different scatterplots
# mRNA vs Deconvolution: 0.079 pearson correlation
ggplot(correlation, aes(x = mRNA, y = Deconvolution)) + 
  geom_point(size = 2) + # scale_color_manual(values = c("red", "green")) +
  geom_text(aes(label = rownames(correlation)), size = 3, nudge_x = 0.25, nudge_y = 0.25, 
            check_overlap = T, color = "black") +
  ylab("MCPCounter Deconvolution median") +
  xlab("RNAseq median expression for DEG") + theme_bw() +
  # facet_wrap(~ Time, scales = "free_x") +
  theme(strip.background = element_rect(fill = "white", color = "black"),
        strip.text.x = element_text(size = 12)) +
  stat_cor(method = "pearson")

# mRNA vs mutation: 0.31 pearson correlation
ggplot(correlation, aes(x = mRNA, y = WES)) + 
  geom_point(size = 2) + # scale_color_manual(values = c("red", "green")) +
  geom_text(aes(label = rownames(correlation)), size = 3, nudge_x = 0.25, nudge_y = 0.25, 
            check_overlap = T, color = "black") +
  ylab("Z-score functional SNV + INDEL mutation data median") +
  xlab("RNAseq median expression for DEG") + theme_bw() +
  # facet_wrap(~ Time, scales = "free_x") +
  theme(strip.background = element_rect(fill = "white", color = "black"),
        strip.text.x = element_text(size = 12)) +
  stat_cor(method = "pearson")

# Deconvolution vs mutation: 0.57 R
ggplot(correlation, aes(x = Deconvolution, y = WES)) + 
  geom_point(size = 2) + # scale_color_manual(values = c("red", "green")) +
  geom_text(aes(label = rownames(correlation)), size = 3, nudge_x = 0.25, nudge_y = 0.25, 
            check_overlap = T, color = "black") +
  ylab("Z-score functional SNV + INDEL mutation data median") +
  xlab("MCPCounter Deconvolution median") + theme_bw() +
  # facet_wrap(~ Time, scales = "free_x") +
  theme(strip.background = element_rect(fill = "white", color = "black"),
        strip.text.x = element_text(size = 12)) +
  stat_cor(method = "pearson")

# List with expression and methylation data sets
data <- list(mRNA = tpm_deg_model_t, WES = zmut_model_t, Deconvolution = zmcp_model_t)

# Design matrix
design <- matrix(0.1, ncol = length(data), nrow = length(data), 
                dimnames = list(names(data), names(data)))
diag(design) <- 0 # matrix diagonal is set to 0 (we do not want correlation between equal omics data)
```

Method may be now applied accounting for the response groups and our design. An initial number of components for the model is selected: in our case, 5 components are initially included. From this result, an evaluation of classification and prediction performance using k-fold cross-validation is performed to estimate the optimal number of components for the model.


```{r}
# Apply method
res <- block.splsda(X = data, Y = cond, ncomp = 5, design = design)

# Performance evaluation
set.seed(123) # for reproducibility, only when the `cpus' argument is not used
# 5-fold cross validation repeated 10 times
perf.diablo <- perf(res, validation = 'Mfold', folds = 5, nrepeat = 10)

plot(perf.diablo)
```

The previous performance plot outputs the classification error rate according to three prediction distances. The minimum point is found at four components by mahalanobis distance. 

Next step is selecting the optimal number of variables in the model for each omics data set. For that, we employ tunning indicating a grid of keepX values that are assessed on each component, one at a time. That is, we use a vector to select different numbers of variables to test. A BPARAM parameter is defined to set the parallelization of the process. For each data set, 5 to 10 features will be tested per component.

```{r}
# Optimal number of components
ncomp = 2

# Tunning
set.seed(123) # for reproducibility, only when the `cpus' argument is not used
test.keepX = list (mRNA = c(5:10),
                   Deconvolution = c(5:10),
                   WES = c(5:10)) 

BPPARAM <- BiocParallel::MulticoreParam(workers = parallel::detectCores()-1)

tune.stad <- tune.block.splsda(X = data, Y = cond, ncomp = ncomp,
                              test.keepX = test.keepX, design = design,
                              validation = 'Mfold', folds = 5, nrepeat = 1,
                              dist = "centroids.dist", BPPARAM = BPPARAM)

# Features selected are listed 
list.keepX <- tune.stad$choice.keepX
list.keepX
```

A number of variables was selected for each of the four components in each layer of information (7 and 7 for mRNA, 6 and 6 for mutations and 5 and 5 for deconvolution).

Now, the final model with the obtained list of features and the number of components may be computed.

```{r}
# Apply method with tunned params
res.tune <- block.splsda(X = data, Y = cond, ncomp = ncomp, 
                          keepX = list.keepX, design = design)

res.tune$design # including our age condition
```

### Plots and results

```{r}
# Individual plots
plotIndiv(res.tune, ind.names = FALSE, 
          legend=TRUE, cex=c(1,2), ellipse = TRUE,
          title = 'Response in individual layers')
```

mRNA acocmplishes a more distinguished differential distribution between groups. 

Next, correlation of omics data can be plotted too.

```{r}
plotDiablo(res.tune, ncomp = 1)
plotDiablo(res.tune, ncomp = 2)
# plotDiablo(res.tune, ncomp = 3)
# plotDiablo(res.tune, ncomp = 4)
```

Correlation is quite high between different omics information.

```{r}
# Arrow sample plot
plotArrow(res.tune, ind.names = FALSE, legend = TRUE, main = 'DIABLO', 
          col.per.group = c("red", "green"), legend.title = "Response")

# Loadings
plotLoadings(res.tune, comp = 1, contrib = 'max', 
             method = 'median', legend.color = c("red", "green"))
plotLoadings(res.tune, comp = 2, contrib = 'max', 
             method = 'median', legend.color = c("red", "green"))
# plotLoadings(res.tune, comp = 3, contrib = 'max', 
#              method = 'median', legend.color = c("red", "green"))
# plotLoadings(res.tune, comp = 4, contrib = 'max', 
#              method = 'median', legend.color = c("red", "green"))

# Correlation circle plot
plotVar(res.tune, var.names = TRUE, style = 'graphics', legend = TRUE, 
        pch = c(16, 17, 18), cex = c(1,1,1))

# Circos plot
circosPlot(res.tune, cutoff = 0.8, size.variables = 0.5, line=TRUE,
           col.per.group = c("red", "green"), Y.name = "Levels")

pdf("/omics/groups/OE0436/data/jimenez/outputs/cimplot_diablo.pdf")

cimDiablo(res.tune, legend.position = "topright", size.legend = 0.6, 
          transpose = TRUE, trim = TRUE, #margins = c(10,15), 
          dist.method = c("correlation", "correlation"), clust.method = c("complete","complete"))

dev.off()

pdf("/omics/groups/OE0436/data/jimenez/outputs/network_diablo.pdf")

network(res.tune, blocks = c(1,2,3),
        color.node = c('blue', 'green', 'red'), 
        cutoff = 0.8)

dev.off()
```

Considering all three layers of information across Dimension 1 and 2, our samples are discriminated by response, as shown above. Thus, we can conclude that integration of data may explain response differences.

### Performance of the model

Finally, we can test the performance of the final model as a classifier plotting a ROC curve. This AUC plot is computed from training cross-validation sets and averaged. Thus, it may give a quick insight into the capability of the model to predict original data. An AUC plot is obtained for each block of omics data using the first two components.

```{r}
# mRNA
auc.mrna <- auroc(res.tune, roc.block = "mRNA", roc.comp = 2)

# Deconvolution
auc.deconvolution <- auroc(res.tune, roc.block = "Deconvolution", roc.comp = 2)

# WES
auc.wes <- auroc(res.tune, roc.block = "WES", roc.comp = 2)
```

Nonetheless, it is important to consider that this prediction is based on our same data. Thus, employing those samples that were not used to build the model, the model can be tested to compute the generalization error.

```{r}
# Prepare test data
# Subset data with not used PIDs
zmcp_test <- zmcp[, -cases_model_idx]
tpm_deg_test <- tpm_deg[, -cases_model_idx]
zmut_test <- zmut[, -cases_model_idx]

# Transpose
zmcp_test_t <- t(zmcp_test)
tpm_deg_test_t <- t(tpm_deg_test)
zmut_test_t <- t(zmut_test)

data.test <- list(mRNA = tpm_deg_test_t, Deconvolution = zmcp_test_t, WES = zmut_test_t)

# Prediction 
predict.diablo = predict(res.tune, newdata = data.test)

# Confussion matrix compares the real and predicted subtypes.
confusion.mat <- get.confusion_matrix(truth = pheno$response[-cases_model_idx], 
                     predicted = predict.diablo$WeightedVote$centroids.dist[,2])
confusion.mat

# ROC curves
auroc(res.tune, roc.block = "Deconvolution", newdata = data.test)

auroc(res.tune, roc.block = "mRNA", newdata = data.test)

auroc(res.tune, roc.block = "WES", newdata = data.test)

# Error rate: incorrect predictions in the total
error_rate <- (confusion.mat[1,2] + confusion.mat[2,1]) / 10
error_rate
```

## References
1.Singh, A. et al. DIABLO: an integrative ap-proach for identifying key molecular drivers from multi-omics assays. doi:10.1093/bioinformatics/bty1054.
2.Rohart, F., Gautier, B., Singh, A. & Lê Cao, K. A. mixOmics: An R package for ‘omics feature selection and multiple data integration. PLoS Comput Biol 13, e1005752 (2017).
3.Tenenhaus, A. et al. Variable selection for gen-eralized canonical correlation analysis. Biosta-tistics 15, 569–583 (2014).
