---
title: <span style="font-family:Georgia;">Conversion of snv and indel vcf files to maf format</span>
author: "Víctor Jiménez Martínez"
date: "2022"
output: 
  html_document:
    toc: true
    theme: united
---

## Introduction

The main goal of this script is to merge and obtain maf files from vcf-like files with SNV and indel information. This script was adapted from the script of Roma Kurilov (22/04/2022) and contains a function for converting all snv and indel .vcf files (generated by DKFZ pipeline) from the project to a single .maf file (which allows to use maftools package). The changes I made affect basically the inf and type variables, which in the original script seem to catch some information from the file name that has not been able to replicate in these files. The identification of the type of modifications in the if loops is done with grepl instead of type variable

## Libraries 

```{r, results="hide", message=FALSE, warning=FALSE}
# R version employed for this script is 4.2.0
library(maftools)
library(dplyr)
```

## Function

```{r}
read_vcfs <- function(files, path){
  maf <- matrix(ncol=39, nrow=0)
  column_list <- c("Hugo_Symbol", "Chromosome", "Start_Position", "End_Position", "Reference_Allele", "Tumor_Seq_Allele2", "Tumor_Sample_Barcode", "Variant_type", "Variant_Classification",
                   "ID", "QUAL", "FILTER", "INFO", "FORMAT", "DBSNP", "X1K_GENOMES", "ANNOVAR_FUNCTION", "ANNOVAR_TRANSCRIPTS", "SEGDUP", "CYTOBAND", "REPEAT_MASKER", "DAC_BLACKLIST", "DUKE_EXCLUDED",         
                   "HISEQDEPTH", "SELFCHAIN", "MAPABILITY", "SIMPLE_TANDEMREPEATS", "CONFIDENCE", "Enhancers", "CpGislands", "TFBScons", "ENCODE_DNASE", "miRNAs_snoRNAs", "miRBase18", "COSMIC", "miRNAtargets",
                   "CgiMountains", "phastConsElem20bp", "ENCODE_TFBS")
  colnames(maf) <- column_list
  
  for (f in files) {
    
    vcf <- read.delim(paste0(path, f))
    # inf <- unlist(strsplit(f, split="/"))
    # Getting the sample IDs from our particular vcf files:
    # If files have a column with path
    # Tumor_Sample_Barcode <- gsub("/icgc/dkfzlsdf/project/hipo2/hipo_K28A/sequencing/exon_sequencing/view-by-pid/.*/snv_results/paired/", "", vcf$FILENAME)
    # Tumor_Sample_Barcode <- gsub("_.*/results_SNVCallingWorkflow-.*/snvs", "", Tumor_Sample_Barcode)
    # Tumor_Sample_Barcode <- gsub("_somatic_functional_snvs_conf_8_to_10.vcf", "", Tumor_Sample_Barcode)
    # If a column with the barcode is already there
    Tumor_Sample_Barcode <- vcf$PID
    # type <- inf[2]
    
    # snvs
    if (grepl("snvs", f)){
      vcf$Variant_type <- rep("SNP", nrow(vcf))
      vcf$Variant_Classification <- recode(vcf$EXONIC_CLASSIFICATION, "nonsynonymous SNV" = "Missense_Mutation", "stopgain"="Nonsense_Mutation", "stoploss"="Nonstop_Mutation")
      vcf$Variant_Classification[which(vcf$Variant_Classification==".")] <- "Splice_Site"
    }
    
    #indels
    if (grepl("indels", f)){
      vcf$EXONIC_CLASSIFICATION[vcf$EXONIC_CLASSIFICATION == "unknown" | vcf$EXONIC_CLASSIFICATION == "."] <- "NA NA"
      vcf$EXONIC_CLASSIFICATION[vcf$EXONIC_CLASSIFICATION == "stopgain"] <- "stop gain"
      temp_mat <- matrix(unlist(strsplit(vcf$EXONIC_CLASSIFICATION, split=" ")), ncol=2, byrow=T)
      colnames(temp_mat) <- c("ex_class1", "ex_class2")
      
      vcf$Variant_type <- rep("INDEL", nrow(vcf)) # toupper(substr(temp_mat[,"ex_class2"], 1, 3))
      vcf$Variant_Classification <- apply(temp_mat, 1, function(x){
        if(x[1]=="frameshift") {res1 <- "Frame_Shift"} else {res1 <- "In_Frame"}
        if(x[2]=="deletion") {res2 <- "Del"} else {res2 <- "Ins"}
        res <- paste0(res1, "_", res2)
        res
      })
    }
    
    # renaming columns
    colnames(vcf)[which(colnames(vcf)=="GENE")] <- "Hugo_Symbol"
    colnames(vcf)[which(colnames(vcf)=="X.CHROM")] <- "Chromosome"
    colnames(vcf)[which(colnames(vcf)=="POS")] <- "Start_Position"
    vcf$End_Position <- vcf$Start_Position
    colnames(vcf)[which(colnames(vcf)=="REF")] <- "Reference_Allele"
    colnames(vcf)[which(colnames(vcf)=="ALT")] <- "Tumor_Seq_Allele2"
    vcf$Tumor_Sample_Barcode <- Tumor_Sample_Barcode
    
    #keeping commmon indel, snv columns
    
    vcf <- vcf[,column_list]
    maf <- rbind(maf, vcf)
  }
  return(maf)
}
```

## Usage

```{r}
path <- "/omics/groups/OE0436/data/jimenez/data/snv_indels/vcf/"
files <- list.files(path = path, recursive = T, pattern = "*metastasis2_definitive.vcf")

maf_test <- read_vcfs(files, path)

## Adding columns that MutScape asks for
maf_test$Transcript_ID <- unlist(lapply(strsplit(maf_test$ANNOVAR_TRANSCRIPTS, ":"), function(x) x[2]))
maf_test$Gene <- maf_test$Hugo_Symbol
maf_test$Protein_position <- unlist(lapply(strsplit(maf_test$ANNOVAR_TRANSCRIPTS, ":"), function(x) x[5]))
maf_test$Protein_position <- unlist(lapply(strsplit(maf_test$Protein_position, ","), function(x) x[1]))
maf_test$Protein_position <- unlist(lapply(strsplit(maf_test$Protein_position, "\\."), function(x) x[2]))
maf_test$Protein_position <- as.numeric(sapply(maf_test$Protein_position, function(x) gsub("([A-Z]+)", "", x)))
maf_test$Protein_position[which(is.na(maf_test$Protein_position))] <- " "

# Subset maf file to create three different maf files per group
pheno <- read.csv(file = "/omics/groups/OE0436/data/jimenez/data/PID_treatment.csv", 
                  header = TRUE, sep = ";", row.names = "PID")
colnames(pheno) = c("therapy", "evolution") # change names

pheno <- pheno %>% 
  mutate(response = ifelse(evolution == "PR", "responders", "non-responders"))

maf_resp <- maf_test[which(maf_test$Tumor_Sample_Barcode %in% rownames(pheno)[pheno$response == "responders"]), ]
maf_non <- maf_test[which(maf_test$Tumor_Sample_Barcode %in% rownames(pheno)[pheno$response == "non-responders"]), ]

# # With only the pre-treatment sample
# maf_PR_pre <- maf_test[which(maf_test$Tumor_Sample_Barcode%in%group$ID[group$response=="PR"] &
#                                maf_test$Tumor_Sample_Barcode%in%group$ID[grepl("metastasis1", group$ID)]), ]
# maf_SD_pre <- maf_test[which(maf_test$Tumor_Sample_Barcode%in%group$ID[group$response=="SD"] &
#                                maf_test$Tumor_Sample_Barcode%in%group$ID[grepl("metastasis1", group$ID)]), ]
# maf_PD_pre <- maf_test[which(maf_test$Tumor_Sample_Barcode%in%group$ID[group$response=="PD"] &
#                                maf_test$Tumor_Sample_Barcode%in%group$ID[grepl("metastasis1", group$ID)]), ]
```

## Maftools functions

```{r}
# It generates some graphs
maf <- read.maf(maf_test)
plotmafSummary(maf = maf)
oncoplot(maf = maf, top = 10)
```

## Save files

```{r}
# save(maf_test, file="snv.maf.tsv")
# write.table(maf_test, file="all_snvs_all_patients.maf", sep="\t", quote=FALSE, row.names=FALSE)
# 
# write.table(maf_PR, file="snv_PR_pre_all_patients.maf", sep="\t", quote=FALSE, row.names=FALSE)
# write.table(maf_SD, file="snv_SD_pre_all_patients.maf", sep="\t", quote=FALSE, row.names=FALSE)
# write.table(maf_PD, file="snv_PD_pre_all_patients.maf", sep="\t", quote=FALSE, row.names=FALSE)

write.table(maf_resp, 
            file = "/omics/groups/OE0436/data/jimenez/data/snv_indels/maf/mut_resp_met2_definitive.maf", 
            sep = "\t", quote = FALSE, row.names = FALSE)
write.table(maf_non, 
            file = "/omics/groups/OE0436/data/jimenez/data/snv_indels/maf/mut_non_met2_definitive.maf",
            sep = "\t", quote = FALSE, row.names = FALSE)
```






